### DISCUSSION TITLE
Prevent downstream mutation of specified record properties

### DISCUSSION THREAD

**Discussion by WhitWaldo** [dotnet/csharplang#7129] [Language Ideas] [answered]
Given an existing record today, I can with the `with` expression and specify some number of the record's properties to change their values on a new record instance and this has already saved me a lot of time and effort.

However, I'm working on receiving and modifying values from an API and it occurs to me that this could be made even more helpful if library authors were able to put restrictions on precisely which properties of a record were able to be changed in a downstream record - sort of like a readonly modifier for the original record, but which is only applied when someone attempts to create a new record using the `with` expression. In other words, there's nothing stopping them from setting these "readonly" properties when creating a brand new record traditionally (e.g. via the constructor, property assignment, etc.).

In this case, when receiving an object from the API, I would like for the downstream user to be able to change any property of the retrieved value except for the createdAt timestamp and the resource identifier, as the server will block any attempt to modify the createdAt value and won't know what to do with the resource if it's provided a modified ID in an update request for a resource that doesn't already exist. Sure, one could make the argument that this server validation is required either way and I agree, but the takeaway here is that as the library author, I could help the developer use my library to avoid this error by keeping them from making the changes to the properties in the first place when otherwise modifying retrieved records.

Offhand, I can't think of a good keyword to use here (fixed, locked, readonly are all taken) , so I'll use "readonly" here as a placeholder, but that wouldn't be the final word used as it has a meaning in C# slightly different than my intention here. But here's an example of what I've got in mind:

```cs
public record Resource(readonly Guid Id, string Name)
{
  public readonly DateTime CreatedAt { get; init; } = DateTime.UtcNow;
}

public void UpdateResource(Resource originalResource)
{
  var updatedResource = originalResource with {
    Name = "UpdatedName"
  }; //Works as expected, leaving Id and CreatedAt unchanged from the original record

  var failedUpdatedResource = originalResource with {
    Name = "UpdatedName",
    CreatedAt = DateTime.UtcNow
  }; //Should fail at development/compile time because the CreatedAt property cannot be modified via a `with` expression
}
```

Thank you for the consideration!

**Comment by iam3yal** [comment]
@WhitWaldo Wouldn't writing an analyzer solve the problem for you?

**Comment by HaloFour** [comment, marked as answer]
You can do this by making the `CreatedAt` property not have an `init` accessor, which is what allows it to be assigned during creation or copying:

```cs
using System;

var bill = new Person(DateTime.UtcNow) {
    Name = "Bill"
};

var frank = bill with {
    Name = "Frank",
    CreatedAt = DateTime.Now // error CS0200: Property or indexer 'Person.CreatedAt' cannot be assigned to -- it is read only
};

public record Person(DateTime createdAt) {
    public DateTime CreatedAt { get; } = createdAt;
    public required string Name { get; init; }
}
```

**Reply by HaloFour** [reply]
Or instead of removing the `init` accessor altogether you could change it's accessibility so that it's not callable outside of your library.  Then you can continue to use initializer syntax but consumers of your library won't be able to.

**Reply by WhitWaldo** [reply]
Thank you for both suggestions!

**Reply by FaustVX** [reply]
The downside is that even if `CreatedAt` property is readonly, there is still `createdAt` property
[sharplab.io](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABATARgFgAoEsAQygAIAjGAG3qoF4qA7CAdyoAVoBnAPZsAFABFywCABUYAWwgA6AKrAAxgDlBnAJRUA3iSrGqG8gpZUARACEG9KyQC+AbhJlKVOFHJsA1pZ0jFScMMAAFgZGJmYWrFYAYj7+VkjRxmpQEJIQKACCwJYSUrIKilrcAPSVVNBQgtQAwgDKAAy4ra0gvPUADtDACFQNVDBsKBAAHtBUAOR8UEJsio1ZOfnAs1RqvmyChTQQVOT8/DAA5hwoVMCCVAC096OFMPxUa9fC9AjObqTE2AAzO8IGoGtcFktxDlSkdMtkpBs9IZiCYqECqMUZPIjqsEbkCgYqOcIMAXFQnJZ4esCn80RisgBHACuMCy12w+FapnMR30xNJ5LGYXJTmcQA==)
this compile
```C#
var frank = bill with {
    Name = "Frank",
    createdAt = DateTime.Now,
};
```

**Reply by HaloFour** [reply]
Oops, yes, change the primary constructor parameter name to `CreatedAt` to match the property and you should be fine.
